### PR TITLE
Add GlamorousToolkit package

### PR DESCRIPTION
**Pull request by akgrant43**
###### Description of changes

Add a package for the GlamorousToolkit, a Smalltalk based, moldable development environment.

See <https://gtoolkit.com/>

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/tools/glamoroustoolkit/default.nix
Comment (AndersonTorres) [inline pkgs/development/tools/glamoroustoolkit/default.nix:1]: Never use pkgs or xorg

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/development/tools/glamoroustoolkit/default.nix
Comment (akgrant43) [reply pkgs/development/tools/glamoroustoolkit/default.nix]: Hi @AndersonTorres ,
Thanks for your feedback.  I've modified it as you suggested.
Cheers,
Alistair

File: pkgs/development/tools/glamoroustoolkit/default.nix
Comment (AndersonTorres) [inline pkgs/development/tools/glamoroustoolkit/default.nix:16]:
```suggestion
{ lib
, stdenv
, fetchzip
, cairo
, dbus
, fontconfig
, freetype
, glib
, gtk3
, libX11
, libXcursor
, libXext
, libXi 
, libXrandr
, libXrender
, libglvnd
, libuuid
, libxcb
, makeWrapper
}:

```

File: pkgs/development/tools/glamoroustoolkit/default.nix
Comment (AndersonTorres) [inline pkgs/development/tools/glamoroustoolkit/default.nix:19]:
```suggestion
  pname = "glamoroustoolkit";
  version = "0.7.2";
```

File: pkgs/development/tools/glamoroustoolkit/default.nix
Comment (AndersonTorres) [inline pkgs/development/tools/glamoroustoolkit/default.nix:24]:
```suggestion
    url = "https://github.com/feenkcom/gtoolkit-vm/releases/download/v${finalAttrs.version}/GlamorousToolkit-x86_64-unknown-linux-gnu.zip";
    stripRoot = false;
    hash = "sha256-Tahe/P0sxmuhSG5JxOkyQGLTOyVrjMEkxvzPc2CBV0Y=";
```

File: pkgs/development/tools/glamoroustoolkit/default.nix
Comment (AndersonTorres) [inline pkgs/development/tools/glamoroustoolkit/default.nix:26]:
```suggestion

  sourceRoot = ".";
```

File: pkgs/development/tools/glamoroustoolkit/default.nix
Comment (AndersonTorres) [inline pkgs/development/tools/glamoroustoolkit/default.nix:32]:
```suggestion
  installPhase = ''
    runHook preInstall

```

File: pkgs/development/tools/glamoroustoolkit/default.nix
Comment (AndersonTorres) [inline pkgs/development/tools/glamoroustoolkit/default.nix:38]:
```suggestion

    runHook postInstall
  '';
```

File: pkgs/development/tools/glamoroustoolkit/default.nix
Comment (AndersonTorres) [inline pkgs/development/tools/glamoroustoolkit/default.nix:60]: Debugging?

File: pkgs/development/tools/glamoroustoolkit/default.nix
Comment (AndersonTorres) [inline pkgs/development/tools/glamoroustoolkit/default.nix:59]:
```suggestion
preFixup = let
    libPath = lib.makeLibraryPath [
      cairo
      dbus
      fontconfig
      freetype
      glib
      gtk3
      libX11
      libXcursor
      libXext
      libXi
      libXrandr
      libXrender
      libglvnd
      libuuid
      libxcb
      stdenv.cc.cc.lib
    ];
  in ''
```

File: pkgs/development/tools/glamoroustoolkit/default.nix
Comment (AndersonTorres) [inline pkgs/development/tools/glamoroustoolkit/default.nix:17]:
```suggestion
stdenv.mkDerivation (finalAttrs: {
```

File: pkgs/development/tools/glamoroustoolkit/default.nix
Comment (AndersonTorres) [inline pkgs/development/tools/glamoroustoolkit/default.nix:37]: This is strange. Have you tried to create the directory tree first?

```suggestion
    install -d $out/bin $out/lib
    cp -r $src/bin $src/lib $out/
    cp ${./GlamorousToolkit-GetImage} $out/bin/GlamorousToolkit-GetImage
```

File: pkgs/development/tools/glamoroustoolkit/default.nix
Comment (AndersonTorres) [inline pkgs/development/tools/glamoroustoolkit/default.nix:94]:
```suggestion
  meta = {
    homepage = "https://gtoolkit.com";
    description = "The GlamorousToolkit Development Environment";
    license = lib.licenses.mit;
    maintainers = [ lib.maintainers.akgrant43 ];
    platforms = lib.platforms.linux;
  };
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Comment by AndersonTorres** [comment]
Are you editing locally?

File: pkgs/development/tools/glamoroustoolkit/GlamorousToolkit-GetImage
Comment (AndersonTorres) [inline pkgs/development/tools/glamoroustoolkit/GlamorousToolkit-GetImage:21]:
```suggestion
wget -q --show-progress \
  -O gt-image.zip \
  https://github.com/feenkcom/gtoolkit/releases/download/${GTVERSION}/GlamorousToolkit-Linux-x86_64-${GTVERSION}.zip 
```

File: pkgs/development/tools/glamoroustoolkit/default.nix
Comment (akgrant43) [reply pkgs/development/tools/glamoroustoolkit/default.nix:70]: Yes, removed.

**Comment by akgrant43** [comment]
> Are you editing locally?

Yes, I'm not at all familiar with github review tools, and most of the changes make sense, so I've just accepted all the changes and am testing locally (although my time is very sporadic today).

File: pkgs/development/tools/glamoroustoolkit/default.nix
Comment (AndersonTorres) [inline pkgs/development/tools/glamoroustoolkit/default.nix:29]:
```suggestion
    stripRoot = false;
    hash = "sha256-Tahe/P0sxmuhSG5JxOkyQGLTOyVrjMEkxvzPc2CBV0Y=";
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]
Squash the commits into one, except the one updating maintainers list

**Comment by akgrant43** [comment]
Hi @AndersonTorres ,

Thanks again for all your feedback.

> Squash the commits into one, except the one updating maintainers list

I've squashed all the commits from today in to a single commit, however there are merges from master prior to that and the rebase command tries to load over 4000 commits and fails.

Would you prefer I close this PR and start again?

Thanks,
Alistair

**Comment by AndersonTorres** [comment]
Have you tried to rebase onto upstream/master, instead of merge?

**Comment by akgrant43** [comment]
I did an interactive rebase on the branch from which the PR is sourced: https://github.com/akgrant43/nixpkgs/tree/glamoroustoolkit-initial, e.g.

```
$ git status
On branch glamoroustoolkit-initial
Your branch is up to date with 'akgrant43/glamoroustoolkit-initial'.
...

$ git rebase -i HEAD~14
```

So the commit shows a single entry for today: https://github.com/akgrant43/nixpkgs/commits/glamoroustoolkit-initial

I think the issue is that I merged master back in to `glamoroustoolkit-initial` between the initial set of commits from 3 weeks ago and today's work.

**Comment by akgrant43** [comment]
OK, I think I understand what you mean now.  I've done a `git rebase master glamoroustoolkit-initial` which seems to have corrected the order of the commits.

I'll try squashing the remaining commits next (excluding the maintainers list).

**Comment by akgrant43** [comment]
OK, now it is just two commits.  I hope I've got it correct. :-)

**Comment by akgrant43** [comment]
I missed one of your suggestions earlier (sha256 format), it's included now.

File: pkgs/development/tools/glamoroustoolkit/default.nix
Comment (AndersonTorres) [inline pkgs/development/tools/glamoroustoolkit/default.nix:101]:
```suggestion
    sourceProvenance = with lib.sourceTypes; [ binaryNativeCode ];
  };
```

File: pkgs/development/tools/glamoroustoolkit/default.nix
Comment (AndersonTorres) [inline pkgs/development/tools/glamoroustoolkit/default.nix:100]:
```suggestion
    platforms = "x86_64-linux";
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Comment by akgrant43** [comment]
I've integrated and squashed the latest two changes.

**Review by AndersonTorres — APPROVED** [review]

**Comment by akgrant43** [comment]
It looks like the platforms value should be a list, instead of a string, i.e.:

```
    platforms = [ "x86_64-linux" ];
```

You've just approved this, I believe I should commit this change, but please let me know what you think.

File: pkgs/development/tools/glamoroustoolkit/default.nix
Comment (AndersonTorres) [inline pkgs/development/tools/glamoroustoolkit/default.nix:27]:
```suggestion
    url = "https://github.com/feenkcom/gtoolkit-vm/releases/download/v${finalAttrs.version}/GlamorousToolkit-x86_64-unknown-linux-gnu.zip";
```

File: pkgs/development/tools/glamoroustoolkit/GlamorousToolkit-GetImage
Comment (AndersonTorres) [inline pkgs/development/tools/glamoroustoolkit/GlamorousToolkit-GetImage:14]: Is it OK to use a version that not matches the installed version?

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Comment by AndersonTorres** [comment]
Also, fix the problems.

**Comment by AndersonTorres** [comment]
> It looks like the platforms value should be a list, instead of a string, i.e.:
> 
> ```
>     platforms = [ "x86_64-linux" ];
> ```

That's right. I haven't noticed.

> 
> You've just approved this, I believe I should commit this change, but please let me know what you think.

File: pkgs/development/tools/glamoroustoolkit/default.nix
Comment (akgrant43) [reply pkgs/development/tools/glamoroustoolkit/default.nix]: Good pickup, thanks!

File: pkgs/development/tools/glamoroustoolkit/GlamorousToolkit-GetImage
Comment (akgrant43) [reply pkgs/development/tools/glamoroustoolkit/GlamorousToolkit-GetImage:14]: You mean that the VM version doesn't match the image version?

GT is built on smalltalk, typically the VM changes multiple times a year while the image changes multiple times per day.

**Comment by akgrant43** [comment]
The `platforms` issue has been resolved.  Unless I've missed something, I'm not aware of any other issues at the moment.  I built the package in a clean VM to check and it's fine.

File: pkgs/development/tools/glamoroustoolkit/GlamorousToolkit-GetImage
Comment (AndersonTorres) [reply pkgs/development/tools/glamoroustoolkit/GlamorousToolkit-GetImage:14]: Nice!

Comment (akgrant43) [reply pkgs/development/tools/glamoroustoolkit/GlamorousToolkit-GetImage:14]: Thanks. :-)

The summary at the bottom of the page (GitHub conversation page) is saying that you are requesting a change here:



However the conversation is showing as resolved, and there's nothing to commit, so I'm not sure what's required.

Thanks again for all your help.

Comment (AndersonTorres) [reply pkgs/development/tools/glamoroustoolkit/GlamorousToolkit-GetImage:14]: GitHub requires a final review in order to remove these.

**Review by AndersonTorres — APPROVED** [review]

**Review by AndersonTorres — CHANGES_REQUESTED** [review]
Finally, squash the two latest commits into only one.

**Comment by akgrant43** [comment]
> Finally, squash the two latest commits into only one.

Done.

File: pkgs/development/tools/glamoroustoolkit/GlamorousToolkit-GetImage
Comment (AndersonTorres) [inline pkgs/development/tools/glamoroustoolkit/GlamorousToolkit-GetImage:21]: Well, you are already using curl to download page info and discover what image to download next.
Why not to use it again here instead of wget?

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/development/tools/glamoroustoolkit/GlamorousToolkit-GetImage
Comment (akgrant43) [reply pkgs/development/tools/glamoroustoolkit/GlamorousToolkit-GetImage]: No reason :-)
Changed, squashed and pushed.

**Review by AndersonTorres — APPROVED** [review]
Let's wait one day to merge.

**Comment by akgrant43** [comment]
Thanks!

**Comment by akgrant43** [comment]
> Let's wait one day to merge.

I was on holiday for 3 weeks and then travelling for work.  I think one day has passed. :-)

Is there anything else I need to do for this to be merged?

Thanks!

**Comment by AndersonTorres** [comment]
Fix conflicts

**Comment by akgrant43** [comment]
Hi @AndersonTorres,

The merge conflict is resolved.  Please let me know if there's anything else I should do.

Thanks!

**Comment by akgrant43** [comment]
Thanks again for all your help, @AndersonTorres !

**Pull request merged**
